testdata/substitutions.go

```diff
@@ -0,0 +1,104 @@
+package testdata
+
+import (
+	"encoding/json"
+	"errors"
+	"strings"
+
+	"gopkg.in/launchdarkly/go-sdk-common.v2/ldvalue"
+)
+
+type substitutionSet map[string]ldvalue.Value
+
+func expandSubstitutions(originalData []byte) ([]SourceInfo, error) {
+	var substs struct {
+		Constants  substitutionSet   `json:"constants"`
+		Parameters []json.RawMessage `json:"parameters"`
+	}
+	if err := ParseJSONOrYAML(originalData, &substs); err != nil {
+		return nil, err
+	}
+	if len(substs.Constants) == 0 && len(substs.Parameters) == 0 {
+		return []SourceInfo{
+			{Data: originalData},
+		}, nil
+	}
+	parameterSets, err := makeParameterPermutations(substs.Parameters)
+	if err != nil {
+		return nil, err
+	}
+	if len(parameterSets) == 0 {
+		return []SourceInfo{
+			{Data: replaceVariables(originalData, substs.Constants)},
+		}, nil
+	}
+	ret := make([]SourceInfo, 0, len(parameterSets))
+	for _, paramsSet := range parameterSets {
+		transformed := replaceVariables(originalData, substs.Constants)
+		transformed = replaceVariables(transformed, paramsSet)
+		transformed = replaceVariables(transformed, substs.Constants)
+		ret = append(ret, SourceInfo{Data: transformed, Params: paramsSet})
+	}
+	return ret, nil
+}
+
+func makeParameterPermutations(paramsData []json.RawMessage) ([]substitutionSet, error) {
+	if len(paramsData) == 0 {
+		return nil, nil
+	}
+	allData, _ := json.Marshal(paramsData)
+	if ldvalue.Parse(paramsData[0]).Type() == ldvalue.ObjectType {
+		var list []substitutionSet
+		if err := json.Unmarshal(allData, &list); err != nil {
+			return nil, err
+		}
+		return list, nil
+	}
+	if ldvalue.Parse(paramsData[0]).Type() != ldvalue.ArrayType {
+		return nil, errors.New("unable to parse parameters - must be an array of objects or an array of arrays")
+	}
+	var lists [][]substitutionSet
+	if err := json.Unmarshal(allData, &lists); err != nil {
+		return nil, err
+	}
+	indices := make([]int, len(lists))
+	var result []substitutionSet
+	for {
+		mergedSet := make(substitutionSet)
+		for i := 0; i < len(lists); i++ {
+			thisSet := lists[i][indices[i]]
+			for k, v := range thisSet {
+				mergedSet[k] = v
+			}
+		}
+		result = append(result, mergedSet)
+		incrementPos := 0
+		for incrementPos < len(lists) {
+			indices[incrementPos]++
+			if indices[incrementPos] < len(lists[incrementPos]) {
+				break
+			}
+			indices[incrementPos] = 0
+			incrementPos++
+		}
+		if incrementPos == len(lists) {
+			return result, nil
+		}
+	}
+}
+
+func replaceVariables(originalData []byte, substs substitutionSet) []byte {
+	str := string(originalData)
+	str = strings.ReplaceAll(str, `\u003c`, "<")
+	str = strings.ReplaceAll(str, `\u003e`, ">")
+	for name, value := range substs {
```

Comment:
This looks, on the surface, like it could become a performance issue depending on the amount of substitution sets & number of substitutions within them, since we'll be calling `strings.ReplaceAll` approximately that many times. 

I'd be curious to measure the current speed and see how it behaves.

Reply:
I don't see that as a significant concern. I mean, yes, it is a performance issue in the sense that it affects performance, and doing this is slower than not doing it. But we're doing it because it's useful, so the question is whether in practice it actually makes the tests slower than we want them to be. And in a compiled Go application, I can't imagine any case in which the speed of doing a bunch of in-memory string replacements, on strings that are just a couple of K in length, could have any noticeable effect on execution time unless we were running hundreds of thousands of tests.

I also can't think how we could profile this in any meaningful way, given that the tests also involve I/O (HTTP communications with the test service) which is always going to be _much_ slower than any kind of string manipulations.

Reply:
I mean, if you run the tests right now, you will see that it's running through the evaluation tests (which are the only ones that currently use data files like this) basically as fast as it can print the output. That's not surprising since, again, the time taken for I/O (both the HTTP, and writing to the console) outweighs nearly everything else.